Docs/DafnyRef/DafnyRef.mdk

```diff
@@ -3727,34 +3755,83 @@ Note that when there is no body, Dafny assumes that the *ensures*
 than these restrictions, there is no semantic difference between using
-ordinary initialization methods and using constructors.
+ordinary initialization methods and using constructors. Classes may
+declare no constructors or one or more constructors.
+
+#### Classes with no explicit constructors
+
+A class that declares no constructors has a default constructor created 
+for it. This constructor is called with the syntax
+```
+C C := new C;
+```
+This constructor simply initializes the fields of the class.
+The order of initialization is such that any field whose value depends on the
+value of another field is initialized after that other field. 
+It is an error if there are circular dependencies. 
+
+This constructor sets each class field to an arbitrary value
+of the field's type if the field declaration has no initializer 
+and to the value of the initializer expression if it does declare an initializer. 
+For the purposes of proving Dafny programs
+correct, assigning an arbitrary initial value means that the program must 
+be correct for any initial value. Compiled, executable versions of the program
+may use a specific initial value 
+(for example, but not necessarily, a zero-equivalent value).
+
+#### Classes with one or more constructors
 
-The Dafny design allows the constructors to be named, which promotes
-using names like `InitFromList` above.  Still, many classes have just
+When one or more constructors are explicitly declared, they are named, 
+which promotes
+using names like `InitFromList` above.
+Constructors must have distinct names, even if their signatures are different.
+Many classes have just
 one constructor or have a typical constructor.  Therefore, Dafny
 allows one _anonymous constructor_, that is, a constructor whose name
 is essentially "".  For example:
 ```
 class Item {
+  constructor I(xy: int) // ...
   constructor (x: int, y: int)
   // ...
 }
 ```
-When invoking this constructor, the "`.`" is dropped, as in:
+The named constructor is invoked as
+```
+  Item i := new Item.I(42);
+```
+The anonymous constructor is invoked as
 ```
-m := new Item(45, 29);
+  Item i := new Item(42,43);
 ```
-Note that an anonymous constructor is just one way to name a
-constructor; there can be other constructors as well.
+dropping the "`.`".
+
+The body of a constructor contains two sections, 
+an initialization phase and a post-initialization phase, separated by a `new;` statement.
+If there is no `new;` statement, the entire body is the initialization phase. 
+This initialization phase is intended to initialize field variables. 
+In this phase, uses of the object reference `this` are restricted;
+a program may use `this`
+
+ - as the receiver on the LHS, 
+ - as the entire RHS of an assignment to a field of `this`, 
+ - and as a member of a set on the RHS that is being assigned to a field of `this`.
+
+Furthermore, `const` fields may only be assigned to (once) in an initialization phase
```

Comment:
You can actually assign to a `const` multiple times in the initialization phase.